File: leetcode/039_combinationsum.py
class Solution(object):
    def backtrack(self, candidates, idx, currlist, currsum, target, result):
        if currsum == target:
            result.append(currlist)
            return
        if currsum > target:
            return
        for i in range(idx,len(candidates)):
            self.backtrack(candidates, i, currlist + [candidates[i]], currsum+candidates[i], target, result)
    
    
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        self.backtrack(candidates, 0, [], 0, target, result)
        return result

File: leetcode/039_combinationsum.ts
function addMore(currNums: number[], currSum: number, idx: number, candidates: number[], target: number, results: number[][]) {
    if (currSum > target) return
    if (currSum === target) {
        results.push(currNums)
    }
    for (let i = idx; i < candidates.length; i++) {
        addMore([candidates[i], ...currNums], currSum + candidates[i], i, candidates, target, results)
    }
}

function combinationSum(candidates: number[], target: number): number[][] {
    const allCombos: number[][] = []
    addMore([], 0, 0, candidates, target, allCombos)
    return allCombos
};
